contracts/RMM01Portfolio.sol

```diff
@@ -28,34 +29,58 @@ contract RMM01Portfolio is PortfolioVirtual {
      * movement.
      * @custom:reverts Underflows if current reserves of output token is less then next reserves.
      */
-    function _getLatestInvariantAndVirtualPrice(uint64 poolId)
+    function _getLatestInvariantAndVirtualPrice(
+        uint64 poolId,
+        bool sellAsset
+    )
         internal
         view
         returns (uint256 price, int256 invariant, uint256 updatedTau)
     {
         PortfolioPool storage pool = pools[poolId];
         updatedTau = pool.computeTau(block.timestamp);
 
-        (uint256 x, uint256 y) = pool.getVirtualPoolReservesPerLiquidityInWad();
+        uint256 reserveXPerLiquidity = pool.virtualX;
+        uint256 reserveYPerLiquidity = pool.virtualY;
+
+        // The reserve which sends tokens out in this swap must be rounded up
+        // to avoid the scenario that the remainder in a truncated output amount
+        // is not stolen in the swap.
+        if (sellAsset) {
+            // Swap X -> Y, Y is output, so round up Y.
+            reserveXPerLiquidity =
+                reserveXPerLiquidity.divWadDown(pool.liquidity);
+            reserveYPerLiquidity = reserveYPerLiquidity.divWadUp(pool.liquidity);
+        } else {
+            // Swap Y -> X, X is output, so round up X.
+            reserveXPerLiquidity = reserveXPerLiquidity.divWadUp(pool.liquidity);
+            reserveYPerLiquidity =
+                reserveYPerLiquidity.divWadDown(pool.liquidity);
+        }
```

Comment:
Now computes the reserves per liquidity amounts using the proper rounding direction depending on the output reserve.

Consequence: output reserve is rounded up to prevent the case that the output reserve PER LIQUIDITY does not change, but the total output reserve does change, leading to some potential to drain the small fractional amount.